sky/provision/gcp/config.py

```diff
@@ -0,0 +1,720 @@
+                if port_list == []:
+                    port_set.update(set(range(1, 65536)))
+                else:
+                    for port_range in port_list:
+                        parse_ports = port_range.split('-')
+                        if len(parse_ports) == 1:
+                            port_set.add(int(parse_ports[0]))
+                        else:
+                            assert (
+                                len(parse_ports) == 2
+                            ), f'Failed to parse the port range: {port_range}'
+                            port_set.update(
+                                set(
+                                    range(int(parse_ports[0]),
+                                          int(parse_ports[1]) + 1)))
+                if allowed['IPProtocol'] not in source2rules[direction_source]:
+                    source2rules[direction_source][
+                        allowed['IPProtocol']] = set()
+                source2rules[direction_source][allowed['IPProtocol']].update(
+                    port_set)
+        return source2rules
+
+    effective_rules = _merge_and_refine_rule(effective_rules)
+    required_rules = _merge_and_refine_rule(required_rules)
+
+    for direction_source, allowed_req in required_rules.items():
+        if direction_source not in effective_rules:
+            return False
+        allowed_eff = effective_rules[direction_source]
+        # Special case: 'all' means allowing all traffic
+        if 'all' in allowed_eff:
+            continue
+        # Check if the required ports are a subset of the effective ports
+        for protocol, ports_req in allowed_req.items():
+            ports_eff = allowed_eff.get(protocol, set())
+            if not ports_req.issubset(ports_eff):
+                return False
+    return True
+
+
+def _create_rules(project_id: str, compute, rules, VPC_NAME):
+    opertaions = []
+    for rule in rules:
+        # Query firewall rule by its name (unique in a project).
+        # If the rule already exists, delete it first.
+        rule_name = rule['name'].format(VPC_NAME=VPC_NAME)
+        rule_list = _list_firewall_rules(project_id,
+                                         compute,
+                                         filter=f'(name={rule_name})')
+        if len(rule_list) > 0:
+            _delete_firewall_rule(project_id, compute, rule_name)
+
+        body = rule.copy()
+        body['name'] = body['name'].format(VPC_NAME=VPC_NAME)
+        body['network'] = body['network'].format(PROJ_ID=project_id,
+                                                 VPC_NAME=VPC_NAME)
+        body['selfLink'] = body['selfLink'].format(PROJ_ID=project_id,
+                                                   VPC_NAME=VPC_NAME)
+        op = compute.firewalls().insert(project=project_id, body=body).execute()
+        opertaions.append(op)
+    for op in opertaions:
+        wait_for_compute_global_operation(project_id, op, compute)
```

Comment:
Seems we have a bunch of duplicated codes here and https://github.com/skypilot-org/skypilot/blob/23aa6dd3f6daf49544d4ddd68acbce422c9d3d03/sky/provision/gcp/instance_utils.py#L497-L539

Can we combine these firewall creations? @cblmemo for a double check.

Reply:
Yeah, I believe they could be merged. Maybe we could move all firewall rule related functions to the `GCPComputeInstance` class?

Reply:
they are somehow different. the one in instance_utils is for managing ports, the code in `config` works for both CPU and TPU instances in the original code. currently I do not have a good way to merge them together

Reply:
After some discussion, the two pieces of code do not serve the same purpose and the details are different. We will keep both for now.